existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,194 @@
+# Existential vs universal quantification
+
+## Introduction
+This article gives brief information about one of the most common extension in Haskell projects - `ExplicitForAll`
+and syntax constructions that this extension brings to language.
+Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints and more!
+
+From type theory we know about the existence of two quantifiers - `$\forall$` and `$\exists$`.
+The former means: `"For every term x the statement is true"`, the later: `"There is at least one term x for which this statement is true"`.
+
+Since Haskell is based on type theory there is a place for these quantifiers in language syntax, however, in Haskell we are talking not about some terms and statements, but about types.
+By default, all type variables are bound by a `forall` quantifier, which means that every type, that passes constraints, can be used inside this function/type declaration.
+
+```haskell
+id :: forall a. a -> a
+id x = x
+```
+
+You can't see this quantification without using a specific extension - "Explicit Forall".
+
+## Universal quantification and ExplicitForall
+By default, `forall.` is implicit and sometimes it can lead to unpleasant situations.
+Consider this code:
+
+```Haskell
+prepend2 :: a -> [a] -> [a]
+prepend2 x xs = pairFun x ++ xs
+where pairFun y = [y, y]
+```
+This code compiles without any errors, until you try to use variables from the outer function inside the `pairFun` function.
+```Haskell
+-- code compiles with error
+prepend2 :: a -> [a] -> [a]
+prepend2 x xs = pair ++ xs
+where pair :: [a]
+      pair = [x, x]
+```
+
+If you try to compile this one you will get an error because compiler can't match type variable `a` from outer
+declaration, with the one in inner declaration.
+This is happening, because the real signature of this function is this:
+
+```Haskell
+prepend2 :: forall a. a -> [a] -> [a]
+prepend2 x xs = pair ++ xs
+where pair :: forall a. [a]
+      pair = [x, x]
+```
+
+In order to correct the code you need to use `-XScopedTypeVariables` extension, which will only work if you
```

Comment:
In this article, we use the `-X` flag prefix for the `ScopedTypeVariables` extension, but not for other extensions.

I think we should be consistent and not use `-X` anywhere.